lib/resin-lint.ts

```diff
@@ -0,0 +1,219 @@
+		extensions: ['ts', 'tsx'],
+		lang: 'typescript',
+	},
+};
+
+/**
+ * The linter expects the path to actual source files, for example:
+ *   src/
+ *   test/
+ * but depcheck expects the root of a project directory (where the
+ * package.json is). This function takes a path and propagates upwards
+ * until it contains a package.json
+ */
+const getPackageJsonDir = function(dir: string): string {
+	let name = findFile('package.json', dir);
+	if (name === null) {
+		throw new Error('Could not find package.json!');
+	}
+	return path.dirname(name);
+};
+
+const read = function(path: string): string {
+	let realPath = fs.realpathSync(path);
+	return fs.readFileSync(realPath).toString();
+};
+
+const findFile = function(name: string, dir?: string): string | null {
+	dir = dir || process.cwd();
+	let filename = path.join(dir, name);
+	let parent = path.dirname(dir);
+	if (fs.existsSync(filename)) {
+		return filename;
+	} else if (dir === parent) {
+		return null;
+	} else {
+		return findFile(name, parent);
+	}
+};
+
+const parseJSON = function(file: string): string {
+	try {
+		return JSON.parse(fs.readFileSync(file).toString());
+	} catch (err) {
+		console.error(`Could not parse ${file}`);
+		throw err;
+	}
+};
+
+const findFiles = function(extensions: string[], paths: string[] = []): string[] {
+	let files = [];
+	for (const p of paths) {
+		if (fs.statSync(p).isDirectory()) {
+			files = files.concat(glob.sync(`${p}/**/*.@(${extensions.join('|')})`));
+		} else {
+			files.push(p);
+		}
+	}
+
+	return files.map(p => path.join(p));
+};
+
+const lintCoffeeFiles = function(files: string[], config: {}): number {
+	const errorReport = new coffeelint.getErrorReport();
+
+	for (const file of files) {
+		const source = read(file);
+		errorReport.lint(file, source, config);
+	}
+
+	const report = new reporter(errorReport, {
+		colorize: process.stdout.isTTY,
+		quiet: false,
+	});
+
+	report.publish();
+
+	return  errorReport.getExitCode();
+};
+
+const lintTsFiles = function(files: string[], config: {}): number {
+	const parsedConfig = tslint.Configuration.parseConfigFile(config);
+	const linter = new tslint.Linter({
+		fix: false,
+		formatter: 'stylish',
+	});
+
+	for (const file of files) {
+		const source = read(file);
+		linter.lint(file, source, parsedConfig);
+	}
+
+	const errorReport = linter.getResult();
+
+	// Print the linter results
+	console.log(linter.getResult().output);
+
+	return errorReport.errorCount === 0 ? 0 : 1;
+};
+
+const runLint = function(resinLintConfig: ResinLintConfig, paths: string[], config: {}) {
+	let linterExitCode;
+	const scripts = findFiles(resinLintConfig.extensions, paths);
+
+	if (resinLintConfig.lang === 'typescript') {
+		linterExitCode = lintTsFiles(scripts, config);
+	}
+
+	if (resinLintConfig.lang === 'coffeescript') {
+		linterExitCode = lintCoffeeFiles(scripts, config);
+	}
+
+	return process.on('exit', () => process.exit(linterExitCode));
+};
+
+export const lint = function(passedParams) {
```

Comment:
Should we use default exports here to keep it closer to the original implementation?

Reply:
You make a good point. We need to check that the call signature in the transpiled code is the same, otherwise this is a breaking change. If it is a breaking change we should probably bundle any other modifications we want at the same time.

Reply:
I don't really mind either export format, but I think you can ignore the breaking change concerns there. This isn't breaking because:

* Changes are only breaking if they break the _declared public API_. I can't see anywhere that documents that this function is exported (only that the binary runner is available).
* We can guarantee that nobody depends on this behaviour. I know they don't, because it won't work in the current release or any releases for a long time. `main` in package.json points to `build/resin-lint.js`, but we quietly stopped publishing that file, or any built JS output, in March 2016: https://github.com/resin-io/node-resin-lint/commit/6026f9fc8daa8a28e8fc5691a815e1a1860a2d71. `require('resin-lint')` hasn't worked for nearly two years.

Feel free to change with impunity, nothing outside this package depends on the internals here :smiley:

Reply:
Oh fantastic...